src/python/pants/core/util_rules/adhoc_process_support.py

```diff
@@ -460,6 +465,95 @@ async def create_tool_runner(
     )
 
 
+async def check_outputs(
+    output_digest: Digest,
+    output_files: Iterable[str],
+    output_directories: Iterable[str],
+    outputs_match_error_behavior: GlobMatchErrorBehavior,
+    outputs_match_mode: GlobExpansionConjunction | None,
+    description_of_origin: str,
+) -> None:
+    if outputs_match_mode is None:
+        return
+
+    filtered_output_files_digests = await MultiGet(
+        Get(
+            Digest,
+            DigestSubset(
+                output_digest,
+                PathGlobs(
+                    [output_file],
+                    glob_match_error_behavior=GlobMatchErrorBehavior.ignore,
+                ),
+            ),
+        )
+        for output_file in output_files
+    )
+
+    filtered_output_directory_digests = await MultiGet(
+        Get(
+            Digest,
+            DigestSubset(
+                output_digest,
+                PathGlobs(
+                    [os.path.join(output_directory, "**")],
+                    glob_match_error_behavior=GlobMatchErrorBehavior.ignore,
+                ),
+            ),
+        )
+        for output_directory in output_directories
+    )
+
+    filtered_output_files = tuple(zip(output_files, filtered_output_files_digests))
+    filtered_output_directories = tuple(zip(output_directories, filtered_output_directory_digests))
+
+    unused_output_files = tuple(
+        f"{output_file} (from Output_files)"
+        for output_file, digest in filtered_output_files
+        if digest == EMPTY_DIGEST
+    )
+    unused_output_directories = tuple(
+        f"{output_directory} (from output_directories)"
+        for output_directory, digest in filtered_output_directories
+        if digest == EMPTY_DIGEST
+    )
+
+    def warn_or_raise(message: str, snapshot: Snapshot) -> None:
+        unused_globs_str = ", ".join([*unused_output_files, *unused_output_directories])
+        message = f"{message}\n\nThe following output globs were unused: {unused_globs_str}"
+        if snapshot.dirs:
+            message += f"\n\nDirectories in output: {', '.join(snapshot.dirs)}"
+        if snapshot.files:
+            message += f"\n\nFiles in output: {', '.join(snapshot.files)}"
```

Comment:
These lists of files and directories are potentially very long (e.g. if one is using it as a wrapper for `npm ci`, I have a relatively small project that ends up with 990 directories and 8299 files). I think we should limit how much we display.

Some ideas:

- just a naive "show the first 10" in whatever order they are here
- something smarter like showing the first 10 ordered by depth (to show the top-level items first, like `top1/, top2/, top3/, top1/a1/, top2/a2, ...`, instead of `top1/, top1/a1, top1/a1/b1, top1/a1/b1/c1, ...`)

something else?

Reply:
Good ideas. The code now sorts by depth (as represented by number of `/` chars) and trims to first 15 entries for each of files and directories.